Yosemite/YosemiteTests/Stores/ProductCategoryStoreTests.swift

```diff
@@ -0,0 +1,104 @@
+
+    func testRetrieveProductCategoriesReturnsCategoriesUponSuccessfulResponse() throws {
+        // Given a stubed product-categories network response
+        let expectation = self.expectation(description: #function)
+        network.simulateResponse(requestUrlSuffix: "products/categories", filename: "categories-all")
+
+        // When dispatching a `synchronizeProductCategories` action
+        let action = ProductCategoryAction.synchronizeProductCategories(siteID: sampleSiteID,
+                                                                     pageNumber: defaultPageNumber,
+                                                                     pageSize: defaultPageSize) { categories, error in
+            // Then a valid set of categories should be returned
+            guard let categories = categories else {
+                return XCTFail("Categories should not be nil.")
+            }
+            XCTAssertFalse(categories.isEmpty)
+            XCTAssertNil(error)
+            expectation.fulfill()
+        }
+
+        store.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    func testRetrieveProductCategoriesReturnsErrorUponReponseError() {
+        // Given a stubed generic-error network response
+        let expectation = self.expectation(description: #function)
+        network.simulateResponse(requestUrlSuffix: "products/categories", filename: "generic_error")
+
+        // When dispatching a `synchronizeProductCategories` action
+        let action = ProductCategoryAction.synchronizeProductCategories(siteID: sampleSiteID,
+                                                                     pageNumber: defaultPageNumber,
+                                                                     pageSize: defaultPageSize) { categories, error in
+            // Then no categories should be returned
+            XCTAssertNil(categories)
+            XCTAssertNotNil(error)
+            expectation.fulfill()
+        }
+
+        store.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    func testRetrieveProductCategoriesReturnsErrorUponEmptyResponse() {
+        // Given a an empty network response
+        let expectation = self.expectation(description: #function)
+
+        // When dispatching a `synchronizeProductCategories` action
+        let action = ProductCategoryAction.synchronizeProductCategories(siteID: sampleSiteID,
+                                                                     pageNumber: defaultPageNumber,
+                                                                     pageSize: defaultPageSize) { categories, error in
+            // Then no categories should be returned
+            XCTAssertNil(categories)
+            XCTAssertNotNil(error)
+            expectation.fulfill()
+        }
+
+        store.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
```

Comment:
My bad. I meant literally separating the Given-When-Then. Here, the "Then" is inside the "When" and when you read the lines, you have to from top to bottom **and then** go back to the middle to look at the assertions. 

Something like this:

```suggestion
    func testRetrieveProductCategoriesReturnsErrorUponReponseError() {
        // Given a stubed generic-error network response
        let expectation = self.expectation(description: #function)
        network.simulateResponse(requestUrlSuffix: "products/categories", filename: "generic_error")

        // When dispatching a `synchronizeProductCategories` action
        var response: (categories: [ProductCategory]?, error: Error?)
        let action = ProductCategoryAction.synchronizeProductCategories(siteID: sampleSiteID,
                                                                     pageNumber: defaultPageNumber,
                                                                     pageSize: defaultPageSize) { categories, error in
            response = (categories, error)
            expectation.fulfill()
        }
        store.onAction(action)
        wait(for: [expectation], timeout: Constants.expectationTimeout)

        // Then no categories should be returned
        XCTAssertNil(response.categories)
        XCTAssertNotNil(response.error)
    }
```

[Here's an article](https://www.swiftbysundell.com/articles/making-swift-tests-easier-to-debug/) (that I haven't read 😅) about the Given-When-Then pattern. You'd also see that there are [Arrange-Act-Assert](http://wiki.c2.com/?ArrangeActAssert) comments in some of our tests. They're similar.

Reply:
Oh I see, I'll try to change them for the next PR!